### PR TITLE
Add test stage in pipeline build jenkins

### PR DESCRIPTION
**Pull request by sophia-guo**
Add enabled test builds in pipeline build jenkins as testing stage (temporarily
adding as non-blocking)
Temporarily using existing parameters set-up in test builds
Trigger different type of test build sequentially(openjdktest,
systemtest)

Signed-off-by: Sophia Guo <sophiag@ca.ibm.com>

### CODE REVIEW

**Comment by gibfahn** [comment]
>I wonder i we can extract some common code into a common Groovy file though. Will that work with the Jenkins Groovy interpreter?

I think these are the relevant docs: https://jenkins.io/doc/book/pipeline/shared-libraries/

Haven't tried it myself though, so YMMV.

**Comment by gdams** [comment]
@sophia-guo take a look at https://stackoverflow.com/a/9154553

**Comment by sophia-guo** [comment]
Definitely we should extract common code to a common Groovy file. However we'd like to do it as next step as talked in build channel:

Heads up that for the build pipelines (https://github.com/AdoptOpenJDK/openjdk-build/tree/master/pipelines), we will make changes in stages:
1)  @sophiag will add a testing stage to them (still as non-blocking a.k.a. propagate : false)
2) then we will introduce variables to avoid as much duplication (do as separate PR since that is a bigger change)
3) I will update the actual Pipeline builds in Jenkins to refer to the ones in github (will do this one platform at a time, during less trafficked times to avoid catastrophic disruption in the event unforeseen problems)
4) when happy that all is working as expected, we can start switching testing to blocking/gating deploy

**Comment by sophia-guo** [comment]
Common function we can use load file as what we do in adoptopenjdk tests:
https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/buildenv/jenkins/openjdk8_x86-64_linux

Common properties we'd prefer yml file, which is supported by pipeline utility readYaml step. 

@gibfahn  if those common functions will be used by different projects shared class library will be a good choice. For this it just make it complicate than expected, which also need the administrator permission to configure.

**Comment by sophia-guo** [comment]
The way I read this, all the build jobs execute in parallel and then once they are ALL complete, all the test jobs execute in parallel. Is this desired behaviour?

Yes, this way we have separate build stage and test stage.

**Comment by sophia-guo** [comment]
Also note: be better if openjdk build and openjdk test build has same Naming Convention.
Example:
build                                                    vs.                               test
openjdk8_build_x86-64_linux                                     openjdk8_**hs**_openjdktest_x86-64_linux
openjdk8_**openj9**_build_x86-64_linux                        openjdk8_**j9**_openjdktest_x86-64_linux

arm64 platform sometime calls "aarch64", sometime calls 'arm64' , 
build                                                    vs.                                test
openjdk9_build_**arm64**_linux                                      openjdk9_hs_openjdktest_**aarch64**_linux
-        build job: 'openjdk10_build_arm64_linux'build:

openjdk8_hs_openjdktest_x86-64_linux

**Comment by smlambert** [comment]
Please note: as @sophia-guo restated above, from what we announced in the build channel, we would like to do these changes in stages.  

This PR is for incorporating the test pipelines.  

A separate PR to follow will be the variables/clean-up work.

**Comment by AdamBrousseau** [comment]
>this way we have separate build stage and test stage

You can still separate the stages and not have a "double diamond graph"
Instead of...
```
      Build         Test
      - X -         - Y -
    /       \     /       \
-  0  - X -  0 - 0  - Y -  0
    \       /     \       /
      - X -         - Y -
```
I suggest...
```
     - XY -
   /        \
- 0  - XY -  0
   \        /
     - XY -
```
Something like this...
```
jobs[platform] = {
    stage('build') {
        buildjobMap[platform] = build job: "openjdk10_build_${archOS}"
    }
    if (platform.test) {
        stage('test') {
            buildjobMap[platform] = build job: "openjdk10_test_${archOS}"
        }
    }
}

parallel jobs
```

**Comment by sophia-guo** [comment]
Yes, this is another choice.

**Comment by AdamBrousseau** [comment]
To be clear, I'm not trying to force anyone into anything, just making sure everyone is clear on what the change does and giving my $0.02 :)

**Comment by karianna** [comment]
I'll only add what I believe the user requirements are.  A release build should never be promoted unless it passes the test pipeline.  nightly builds we can be more lenient.  If the pipeline syntax supports that functional requirement then I'm happy.  Non functional requirements like "make the pipeline more efficient" I'll defer to the Jenkins experts :-)

**Comment by sophia-guo** [comment]
I'd like to confirm nightly build promotion story. Currently if one platform sdk build fails none of any other successful sdk will be promoted. Is that expected? See nightly openjdk8 the latest one is March 8.  https://adoptopenjdk.net/nightly.html?variant=openjdk8  or https://api.adoptopenjdk.net/openjdk8/nightly

This is caused by zlinux failure  https://ci.adoptopenjdk.net/job/openjdk8_build_s390x_linux/

**Comment by smlambert** [comment]
We were actually just discussing this, whether we should rethink the nightly deployment strategy...

@karianna @tellison - thoughts on the matter?  

If one platform fails to build, no nightly builds are deployed.  This definitely makes sense from a release build perspective, and maybe even nightly if builds were more stabilized, but for now as noted, the last published nightly for openjdk8 was March 8 (3 weeks ago), and the last openjdk8-openj9 was March 23 (1 week ago).

**Comment by AdamBrousseau** [comment]
If the pipeline was changed to something like I suggested above then each platform could complete independently of one another. I'm not familiar with your release vs nightly builds but you could have a final stage(s) that would only run once all are passed.

```
     - XY -
   /        \
- 0  - XY -  Z
   \        /
     - XY -
```

File: pipelines/openjdk10_nightly_pipeline.groovy [hunk 1]
```diff
@@ -3,31 +3,48 @@
  * please create a PR there before copying the code over
  */
```
[the remark this answers is not in the document]
Comment (karianna) [reply pipelines/openjdk10_nightly_pipeline.groovy]: Correct those comments can go

**Comment by karianna** [comment]
@smlambert I'd argue that nightly builds can go out independently per platform.  I'd even argue that we could do the same for release (If a AIX build fails for example, I don't feel we should stop the other platform users from getting their fix of Java :-)

**Comment by sophia-guo** [comment]
Updated job workflow. Thanks @AdamBrousseau !

Deployment strategy keep the same. If everyone agrees with @smlambert @karianna we can do it in separate PR.

**Comment by smlambert** [comment]
Thanks for adding the missing groovy script @sophia-guo and incorporating some of the changes discussed.  I will plan to merge this PR today and we can proceed to some of the other changes in separate PRs.

**Pull request merged**